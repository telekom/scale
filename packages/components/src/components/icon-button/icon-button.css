/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --width: var(--telekom-spacing-unit-x6);
  --height: var(--telekom-spacing-unit-x6);
  --width-medium: var(--telekom-spacing-unit-x8);
  --height-medium: var(--telekom-spacing-unit-x8);
  --radius: var(--telekom-radius-standard);
  --box-shadow-focus: 0 0 0 var(--telekom-line-weight-highlight)
    var(--telekom-color-functional-focus);
  --background: var(--telekom-color-background-canvas);
  --background-hover: var(--telekom-color-ui-state-fill-hovered);
  --background-pressed: var(--telekom-color-ui-state-fill-pressed);
  --background-active: var(--telekom-color-ui-state-fill-pressed);
  --spacing-icon-x: var(--telekom-spacing-unit-x2);
  --vertical-align: middle;
  --color: var(--telekom-color-text-and-icon-standard);

  display: inline-flex;
}

[part~='base'] {
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  position: relative;
  border: 0;
  border-radius: var(--radius);
  outline: none;
  cursor: pointer;
  user-select: none;
  font-family: inherit;
  word-spacing: inherit;
  letter-spacing: inherit;
  justify-content: center;
  text-decoration: none;
  padding: 0;
  background: var(--background);
  width: var(--width);
  height: var(--height);
  vertical-align: var(--vertical-align);
}

[part~='base'][part~='size-medium'] {
  height: var(--height-medium);
  width: var(--width-medium);
}

[part~='base']:hover {
  background-color: var(--background-hover);
}

[part~='base']:active {
  background-color: var(--background-pressed);
}

[part~='base']:focus {
  box-shadow: var(--box-shadow-focus);
}

[part~='base'][part~='active'] {
  background: var(--background-active);
}

[part~='base'][part~='active']:hover {
  background: var(--background-hover);
}

[part~='base'][part~='active']:active {
  background: var(--background-pressed);
}

[part~='base'][part~='active']:focus {
  background: var(--background-active);
  box-shadow: var(--box-shadow-focus);
}

slot {
  color: var(--color);
}
