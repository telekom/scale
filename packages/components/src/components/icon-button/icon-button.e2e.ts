/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newE2EPage } from '@stencil/core/testing';

describe('scale-icon-button', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent(
      '<scale-icon-button><scale-icon-action-search></scale-icon-action-search></scale-icon-button>'
    );
    const element = await page.find('scale-icon-button');
    expect(element).toHaveClass('hydrated');
  });

  it('should set tabindex on its inner element', async () => {
    const page = await newE2EPage();
    await page.setContent(
      '<scale-icon-button inner-tabindex="5"><scale-icon-action-search></scale-icon-action-search></scale-icon-button>'
    );
    const element = await page.find('scale-icon-button >>> button');
    expect(element.getAttribute('tabindex')).toBe('5');
  });
});
