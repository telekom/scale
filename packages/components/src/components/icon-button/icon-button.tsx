/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, Prop, h, Host, Element, Method } from '@stencil/core';
import classNames from 'classnames';

@Component({
  tag: 'scale-icon-button',
  styleUrl: 'icon-button.css',
  shadow: true,
})
export class IconButton {
  @Element() hostElement: HTMLElement;

  /** (optional) The size of the element */
  @Prop() size?: 'small' | 'medium' = 'medium';
  /** (optional) Button type */
  @Prop() type?: 'reset' | 'submit' | 'button';
  /** (optional) The name of the button, submitted as a pair with the button's `value` as part of the form data */
  @Prop() name?: string;
  /** (optional) Defines the value associated with the button's `name` */
  @Prop() value?: string;
  /** (optional) Name of a file to be downloaded */
  @Prop() download?: string;
  /** (optional) Set `tabindex` in the inner button or link element */
  @Prop() innerTabindex?: number;
  /** (optional) Set the element to active state  */
  @Prop() active?: boolean;

  private focusableElement: HTMLElement;

  @Method()
  async setFocus() {
    this.focusableElement.focus();
  }

  render() {
    const basePart = classNames(
      'base',
      this.size && `size-${this.size}`,
      this.active && `active`
    );

    return (
      <Host>
        <button
          ref={(el) => (this.focusableElement = el)}
          type={this.type}
          part={basePart}
          tabIndex={this.innerTabindex}
          name={this.name}
          value={this.value}
          aria-pressed={this.active ? 'true' : 'false'}
        >
          <slot />
        </button>
      </Host>
    );
  }
}
