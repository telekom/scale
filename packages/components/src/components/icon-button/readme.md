# scale-icon-button



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description                                                                                               | Type                              | Default     |
| --------------- | ---------------- | --------------------------------------------------------------------------------------------------------- | --------------------------------- | ----------- |
| `active`        | `active`         | (optional) Set the element to active state                                                                | `boolean`                         | `undefined` |
| `download`      | `download`       | (optional) Name of a file to be downloaded                                                                | `string`                          | `undefined` |
| `innerTabindex` | `inner-tabindex` | (optional) Set `tabindex` in the inner button or link element                                             | `number`                          | `undefined` |
| `name`          | `name`           | (optional) The name of the button, submitted as a pair with the button's `value` as part of the form data | `string`                          | `undefined` |
| `size`          | `size`           | (optional) The size of the element                                                                        | `"medium" \| "small"`             | `'medium'`  |
| `type`          | `type`           | (optional) Button type                                                                                    | `"button" \| "reset" \| "submit"` | `undefined` |
| `value`         | `value`          | (optional) Defines the value associated with the button's `name`                                          | `string`                          | `undefined` |


## Methods

### `setFocus() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [scale-search-input](../search-input)

### Graph
```mermaid
graph TD;
  scale-search-input --> scale-icon-button
  style scale-icon-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
