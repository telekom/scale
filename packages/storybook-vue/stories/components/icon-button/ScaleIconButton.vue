
<template>
  <scale-icon-button
    :size="size"
    :active="active"
    :innerTabindex="innerTabindex"
    :download="download"
    :value="value"
    :name="name"
    :type="type"
  >
    <slot></slot>
  </scale-icon-button>
</template>

<script>

export default {
  props: {
    size: { type: String },
    type: { type: String },
    name: { type: String },
    value: { type: String },
    download: { type: String },
    innerTabindex: { type: String },
    active: { type: String },
  },
};
</script>
