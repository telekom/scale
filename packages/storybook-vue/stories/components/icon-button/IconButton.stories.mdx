import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs';
import ScaleIconButton from './ScaleIconButton';

<Meta
  title="Beta Components/Icon Button"
  component={ScaleIconButton}
  argTypes={{}}
/>

## Standard

<Canvas withSource="none">
  <Story name="Standard" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="medium">
          <scale-icon-alert-help accessibility-title="help" size="24"></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

## Selected

<Canvas withSource="none">
  <Story name="Selected" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="medium">
          <scale-icon-alert-help accessibility-title="help" size="24" selected></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="medium">
  <scale-icon-alert-help
    accessibility-title="help"
    size="24"
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Active

<Canvas withSource="none">
  <Story name="Active" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="medium" active="true">
          <scale-icon-alert-help accessibility-title="help" size="24"></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="medium" active="true">
  <scale-icon-alert-help
    accessibility-title="help"
    size="24"
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Active Selected

<Canvas withSource="none">
  <Story name="Active Selected" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="medium" active="true">
          <scale-icon-alert-help accessibility-title="help" size="24" selected></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="medium" active="true">
  <scale-icon-alert-help
    accessibility-title="help"
    size="24"
    selected
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Small

<Canvas withSource="none">
  <Story name="Small" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="small">
          <scale-icon-alert-help accessibility-title="help" size="16"></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="small">
  <scale-icon-alert-help
    accessibility-title="help"
    size="16"
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Small Selected

<Canvas withSource="none">
  <Story name="Small Selected" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="small">
          <scale-icon-alert-help accessibility-title="help" size="16" selected></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="small">
  <scale-icon-alert-help
    accessibility-title="help"
    size="16"
    selected
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Small Active

<Canvas withSource="none">
  <Story name="Small Active" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="small" active="true">
          <scale-icon-alert-help accessibility-title="help" size="16"></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="small" active="true">
  <scale-icon-alert-help
    accessibility-title="help"
    size="16"
  ></scale-icon-alert-help>
</scale-icon-button>
```

## Small Active Selected

<Canvas withSource="none">
  <Story name="Small Active Selected" args={{}}>
    {() => ({
      template: `
        <scale-icon-button size="small" active="true">
          <scale-icon-alert-help accessibility-title="help" size="16" selected></scale-icon-alert-help>
        </scale-icon-button>
      `,
    })}
  </Story>
</Canvas>

```html
<scale-icon-button size="small" active="true">
  <scale-icon-alert-help
    accessibility-title="help"
    size="16"
    selected
  ></scale-icon-alert-help>
</scale-icon-button>
```
