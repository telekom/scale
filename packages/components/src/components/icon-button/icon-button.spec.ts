/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { IconButton } from './icon-button';

describe('IconButton', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [IconButton],
      html: `<scale-icon-button><scale-icon-action-search></scale-icon-action-search></scale-icon-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle click', async () => {
    const page = await newSpecPage({
      components: [IconButton],
      html: `<scale-icon-button><scale-icon-action-search></scale-icon-action-search></scale-icon-button>`,
    });
    expect(page.root).toMatchSnapshot();
    simulateMouseEvent(
      page,
      'click',
      'button',
      expect(page.root).toMatchSnapshot()
    );
  });
});

/**
 * - Helper function for the onMouseEvents
 * @param page = The component, in this case the rating star component
 * @param eventType = The MouseEvent that should be executed on the element
 * @param element = Id of the element on which the MouseEvent is to be executed
 * @param callBack = Testing function
 */
async function simulateMouseEvent(
  page: SpecPage,
  eventType: string,
  element: string,
  callBack: any
) {
  const event = new MouseEvent(eventType, {
    view: window,
    bubbles: true,
    cancelable: true,
  });
  const myTarget = page.root.shadowRoot.querySelector(element);
  const canceled = !myTarget.dispatchEvent(event);
  if (canceled) {
    alert('canceled');
  } else {
    alert('not canceled');
    callBack; // tslint:disable-line
  }
}
