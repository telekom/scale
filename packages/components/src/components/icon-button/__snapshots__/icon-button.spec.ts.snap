// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should handle click 1`] = `
<scale-icon-button>
  <mock:shadow-root>
    <button aria-pressed="false" part="base size-medium">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <scale-icon-action-search></scale-icon-action-search>
</scale-icon-button>
`;

exports[`IconButton should handle click 2`] = `
<scale-icon-button>
  <mock:shadow-root>
    <button aria-pressed="false" part="base size-medium">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <scale-icon-action-search></scale-icon-action-search>
</scale-icon-button>
`;

exports[`IconButton should match snapshot 1`] = `
<scale-icon-button>
  <mock:shadow-root>
    <button aria-pressed="false" part="base size-medium">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <scale-icon-action-search></scale-icon-action-search>
</scale-icon-button>
`;
